src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -0,0 +1,503 @@
+                return completion;
+            }
+        }
+
+        /// <summary>
+        /// Set the given exception as the current activation exception. This has no effect if there
+        /// is not a current activation or there is already an activation exception set.
+        /// </summary>
+        /// <param name="exc">Exception to set.</param>
+        public void SetCurrentActivationException(Exception exc) => currentActivationException ??= exc;
+
+        /// <inheritdoc/>
+        protected override IEnumerable<Task>? GetScheduledTasks() => scheduledTasks;
+
+        /// <inheritdoc/>
+        protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
+        {
+            // No benefit for executing inline, so just append to list if not there
+            if (!taskWasPreviouslyQueued)
+            {
+                QueueTask(task);
+            }
+            return false;
+        }
+
+        /// <inheritdoc/>
+        protected override void QueueTask(Task task)
+        {
+            // Only queue if not already done
+            if (!scheduledTaskNodes.ContainsKey(task))
+            {
+                scheduledTaskNodes[task] = scheduledTasks.AddFirst(task);
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override bool TryDequeue(Task task)
+        {
+            // Can't use remove overload that gets value at same time in oldest supported version
+            if (scheduledTaskNodes.TryGetValue(task, out var node))
+            {
+                scheduledTaskNodes.Remove(task);
+                scheduledTasks.Remove(node);
+                return true;
+            }
+            return false;
+        }
+
+        private void RunOnce(bool checkConditions)
+        {
+            // Run as long as we have scheduled tasks
+            // TODO(cretz): Fix to run as long as any tasks not yielded on Temporal
+            while (scheduledTasks.Any())
+            {
+                while (scheduledTasks.Any())
+                {
+                    // Pop last
+                    var task = scheduledTasks.Last!.Value;
+                    scheduledTasks.RemoveLast();
+                    scheduledTaskNodes.Remove(task);
+
+                    // This should never return false
+                    if (!TryExecuteTask(task))
+                    {
+                        Logger.LogWarning("Task unexpectedly was unable to execute");
+                    }
+                    if (currentActivationException != null)
+                    {
+                        ExceptionDispatchInfo.Capture(currentActivationException).Throw();
+                    }
+                }
+
+                // TODO(cretz): Check conditions
+            }
+        }
+
+        private void AddCommand(WorkflowCommand cmd)
+        {
+            if (completion == null)
+            {
+                throw new InvalidOperationException("No completion available");
+            }
```

Comment:
Seems like this state could be made unrepresentable by having completion be passed through the relevant functions instead of as a class field?

Reply:
Correct for this PR, but the next one will have things on outbound interceptor (e.g. starting a timer) that is too far separated from this to have a completion passed through.